### PR TITLE
Clarify use of setters in DistributionExtension; make var/log and var/run excludes configurable

### PR DESCRIPTION
**Pull request by uschi2000**

### CODE REVIEW

**Comment by uschi2000** [comment]
This might be controversial since it's a config-breaking change for anyone taking an upgrade. Rationale: simplify the semantics of the distribution object and make its behaviour more Groovy/Gradle standard. Example:

Old behaviour:

```
distribution {
  args 'foo'
  args 'bar'
}
assert distribution.args == ['bar'] // oops, probably unexpected
```

New behaviour:

```
distribution {
  args = ['foo']
  args = ['bar']
}
assert distribution.args == ['bar'] // expected as per above syntax

distribution {
  args = ['foo']
  args << 'bar' << 'boom'
}
assert distribution.args == ['foo', 'bar', 'boom'] // cool, can use standard groovy things

```

**Comment by schlosna** [comment]
Rob, can you provide the context on this change?

I think you want to add a test case for the last example above as I believe that should include foo and not bar twice:

```
distribution {
  args = ['foo']
  args << 'bar' << 'boom'
}
assert distribution.args == ['foo', 'bar', 'boom'] // cool, can use standard groovy things
```

**Comment by uschi2000** [comment]
thanks, I have a typo in my example. edited/fixed above.

**Comment by uschi2000** [comment]
Context: I consider the current configuration mechanism a bit whacky and would like to remove tech debt.

**Comment by markelliot** [comment]
I'm not in favor of this change -- it's more idiomatic Gradle to use the setters, IMO.

**Comment by uschi2000** [comment]
that's not quite accurate, if I remember right. typically, gradle objects have setters (whose names start with `set`) and "accumulators" (for the lack of a better name). this would look like this:

```
distribution {
  args "foo"
  args "bar"
  // args is now [foo, bar]
  setArgs ["boom", "baz"]
  // args is now [boom, baz]
}
```

happy to support both here, but the current model is a little off IMO

**Comment by markelliot** [comment]
Haven't seen the set method syntax -- this is a poor man's replication of the slightly fancier Action syntax the Gradle core stuff uses: https://github.com/gradle/gradle/blob/master/subprojects/publish/src/main/java/org/gradle/api/publish/PublishingExtension.java

**Comment by uschi2000** [comment]
here's an example from DefaultSourceDirectorySet.

```
    public SourceDirectorySet srcDirs(Object... srcDirs) {
        for (Object srcDir : srcDirs) {
            source.add(srcDir);
        }
        return this;
    }

    public SourceDirectorySet setSrcDirs(Iterable<?> srcPaths) {
        source.clear();
        GUtil.addToCollection(source, srcPaths);
        return this;
    }
```

**Comment by uschi2000** [comment]
will change this pr to adopt that convention.

**Comment by markelliot** [comment]
Are there any examples of single-valued things? Not all of our parameters are iterable.

**Comment by uschi2000** [comment]
yep, standard setter

```
    public void setSourceCompatibility(Object value) {
        srcCompat = JavaVersion.toVersion(value);
    }
```

**Comment by markelliot** [comment]
sourceCompatibility is done using the old container method instead of extensions; I think we want to model what we're doing after extensions.

**Comment by uschi2000** [comment]
rebased. to summarize:

Simplify the semantics of the distribution object and make its behaviour more Groovy/Gradle standard. Example:

Old behaviour:

```
distribution {
  args 'foo'
  args 'bar'
}
assert distribution.args == ['bar'] // oops, probably unexpected
```

New behaviour:

```
distribution {
  args 'foo'
  args 'bar'
}
assert distribution.args == ['foo', 'bar']

distribution {
  args 'foo'
  setArgs(['bar'])
}
assert distribution.args == ['bar']
```

**Comment by markelliot** [comment]
lgtm. I think the `=` signs in your example of new behavior aren't supposed to be there, though.

File: readme.md [hunk 1]
```diff
@@ -65,9 +65,9 @@ The `distribution` block offers the following options:
...
    classpath is omitted from the generated Windows start script and instead inferred
    from a JAR file whose MANIFEST contains the classpath entries.
- * (optional) `excludeFromVar` a list of directories (relative to `${projectDir}/var`) to exclude 
-   in addition to `log` and `run` (which are always excluded).
-   **Note**: this plugin will _always_ create `var/data/tmp` in the resulting distribution to
+ * (optional) `excludeFromVar` a list of directories (relative to `var`) to exclude from the distribution,
+   defaulting to `['log', 'run']`.
+   **Note**: this plugin will *always* create `var/data/tmp` in the resulting distribution to
    ensure the prescribed Java temp directory exists. Setting `data` for this option will still ensure
    nothing in `${projectDir}/var/data` is copied.
```
Comment (markelliot) [inline readme.md:72]: you removed `${projectDir}` above, but not here?

**Pull request merged**